hud/src/androidLibMain/kotlin/HUD.kt

```diff
@@ -149,51 +138,38 @@ class AndroidHUD private constructor(@LayoutRes viewResId: Int, hudConfig: HudCo
     }
 
     private val loadingDialog = LoadingDialog.newInstance(viewResId, hudConfig)
-    private var dialogState = MutableLiveData<DialogState>()
+    private var dialogState = ConflatedBroadcastChannel<DialogState>(DialogState.Gone)
 
     init {
-        checkNotNull(uiContextObserver.uiContextData) { "Please subscribe builder to existing UI context before building HUD" }
-        subscribeIfNeeded(uiContextObserver.uiContextData)
-        uiContextObserver.onUiContextDataWillChange = { newValue, oldValue ->
-            unsubscribeIfNeeded(oldValue)
-            subscribeIfNeeded(newValue)
-        }
-    }
-
-    private fun unsubscribeIfNeeded(uiContextData: UiContextObserver.UiContextData?) {
-        if (uiContextData != null) {
-            runBlocking(Dispatchers.Main.immediate) {
-                dialogState.removeObservers(uiContextData.lifecycleOwner)
-            }
-        }
-    }
-
-    private fun subscribeIfNeeded(uiContextData: UiContextObserver.UiContextData?) {
-        if (uiContextData != null) {
-            runBlocking(Dispatchers.Main.immediate) {
-                dialogState.observe(uiContextData.lifecycleOwner, Observer {
-                    when (it) {
-                        is DialogState.Visible -> loadingDialog.show(uiContextData.fragmentManager, "Kaluga.HUD")
-                        is DialogState.Gone -> loadingDialog.dismiss()
-                    }
-                })
+        launch {
+            combine(uiContextObserver.uiContextData, dialogState.asFlow()) { uiContextData, dialogPresentation ->
+                Pair(uiContextData, dialogPresentation)
+            }.collect { contextPresentation ->
+                when (contextPresentation.second) {
+                    is DialogState.Visible -> contextPresentation.first?.fragmentManager?.let { loadingDialog.show(it, "Kaluga.HUD") }
+                    is DialogState.Gone -> if (loadingDialog.isAdded) loadingDialog.dismiss()
+                }
             }
         }
     }
 
-    override val isVisible get() = loadingDialog.isVisible
+    actual val isVisible get() = loadingDialog.isVisible
 
-    override suspend fun present(animated: Boolean): HUD = suspendCoroutine { continuation ->
-        loadingDialog.presentCompletionBlock = {
-            continuation.resume(this)
+    actual suspend fun present(animated: Boolean): HUD {
+        dialogState.send(DialogState.Visible)
+        return suspendCoroutine { continuation ->
+            loadingDialog.presentCompletionBlock = {
```

Comment:
Can't the send above (and actions triggered by it) complete before the `presentCompletionBlock` is set? The old code only posted after the block was set.